### DISCUSSION TITLE
ViewStore is publishing changes from within view updates

### DISCUSSION THREAD

**Discussion by StefanCosminR** [pointfreeco/swift-composable-architecture#1319] [General]
### Description

I am not sure if this is an Xcode issue because I don't remember seeing this in the past. Reproducing this is very easy, as you can see in the following screenshot. Sending any action using the ViewStore results in this kind of warning.

![Screenshot 2022-08-31 at 15 11 58](https://user-images.githubusercontent.com/17311242/187675727-daa1874e-ab1c-442d-bdc1-f11e550182b5.png)

### Checklist

- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues).

### The Composable Architecture version information

main, protocol-beta, 0.39.1

### Destination operating system

iOS 16

### Xcode version information

Version 14.0 beta 6 (14A5294g)

### Swift Compiler version information

```shell
swift-driver version: 1.62.3 Apple Swift version 5.7 (swiftlang-5.7.0.123.8 clang-1400.0.29.50)
Target: arm64-apple-macosx12.0
```

**Comment by StefanCosminR** [comment]
Here is a sample project 
[ComposableArchitectureWarning.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9461431/ComposableArchitectureWarning.zip)

**Comment by mbrandonw** [comment]
Hi @StephenCosimo, this seems to be a regression in iOS 16 and/or Xcode 14. The equivalent vanilla SwiftUI version of your demo also reproduces the problem:

```swift
import SwiftUI

class Model: ObservableObject {
  @Published var number = 0

  func increment() {
    self.number += 1
  }
}

struct ContentView: View {
  @StateObject var model = Model()

  var body: some View {
    Form {
      Button {
        self.model.increment()
      } label: {
        Text("Increment")
      }
    }
  }
}
```

<img width="863" alt="image" src="https://user-images.githubusercontent.com/135203/187685094-2170fa2d-2387-4059-ab7a-afa40055d422.png">

I'm going to move this to a discussion since it is not a problem with the library per se, but if anyone has any ideas of what the heck is going on with this I'd love to know!

**Reply by tgrapperon** [reply]
This is very likely a beta regression. The issue doesn't happen on Xcode 14b3 but it does with Xcode 14b5. It needs `Form` or `List` to happen (at least). They're likely not telling the system they finished updating. I'm relatively optimistic for this to be fixed in the RC.

**Reply by andreweades** [reply]
https://developer.apple.com/forums/thread/711899

"From my experiments what it means is a scenario like the following:

Views like Sheet(), Picker(), etc. take a binding and change the binding implicitly.
You pass a view model's published property to those views as the binding parameter.
In your view model's code, when that published property changes, other published properties are changed accordingly.

This is the scenario not allowed in beta 5.

So, how to solve the issue? My solution is to introduce a local state to avoid simultaneous changes and use onChange() to sync the local state change with view model. This requires a little bit more code but the error message is gone." - [rayx](https://developer.apple.com/forums/profile/rayx)

**Reply by mbrandonw** [reply]
@andreweades That's an interesting workaround, but I don't really buy the reasoning behind it and do think this is just an Apple bug. First of all, the bug is reproducible without sheets/pickers/etc. Just a simple button in a form will do it as my code sample showed above. And second, there is nothing wrong with writing to bindings even when you hand them off to SwiftUI's native components. It would be bizarre and cumbersome if we needed to maintain local state for every binding we hand to a text field, picker, sheet, etc. I just don't think that's how SwiftUI is meant to be used.

**Reply by iampatbrown** [reply]
@andreweades the workaround would also prevent changes from animating correctly. It seems like a bug that will likely be fixed in the next update.

**Reply by tgrapperon** [reply]
The issue is still present in the RC…
It shows a warning when in a `List` or `Form`, but not when in a `VStack`, so this is extremely likely a bug. 
It seems we'll have to live with it until it's fixed…

**Reply by tgrapperon** [reply]
This issue seems to be fixed with the latest Xcode 14.1b3.
